### PR TITLE
RFC: RoundNearest argument for rem

### PR DESCRIPTION
**Pull request by simonbyrne**
This is my attempt at resolving #9283. It adds an extra argument to `rem`, so that

``` julia
rem(x,y,r::RoundingMode) = x - y*round(x/y,r)
```

without intermediate rounding, so that the result is always in the interval `[-abs(y)/2,abs(y)/2]`. This corresponds to the `remainder` function in the IEEE754 spec.

I've also implemented the corresponding `rem2pi` function, which is more accurate, and seems to be more useful, than `mod2pi` (I actually replaced the only 2 occurrences of `mod2pi` in Base, as this is what they were actually trying to get at).

If we're happy with this, I can add some docs, tests, and implement corresponding `div` functions.

### CODE REVIEW

**Comment by simonbyrne** [comment]
And I guess this would also be useful to have for integer arguments.

**Comment by simonbyrne** [comment]
Any suggestions for why the build is breaking on 32bit?

**Comment by pao** [comment]
Tuple overhaul related. Should be fixed by https://github.com/JuliaLang/julia/commit/7ec501d6e85aa1614c3a330bbbbf907be79bed37. Either force-push or get someone who can rerun the tests to trigger it and you should be good to go.

**Comment by JeffBezanson** [comment]
Looks really good; please continue!

**Comment by StefanKarpinski** [comment]
Yes, this is great. I like the design.

**Comment by StefanKarpinski** [comment]
@simonbyrne – this was generally well-liked, can we get it rebased and merged?

File: base/exports.jl [hunk 1]
```diff
@@ -397,6 +397,8 @@ export
     reim,
     reinterpret,
     rem,
+    rem1,
+    rem2pi,
```
Comment (tkelman) [inline base/exports.jl]: I thought we removed rem1, and does rem2pi need to be exported?

**Comment by StefanKarpinski** [comment]
Ok, I resolved the conflicts as best I could, but to be perfectly honest, I have no idea what I'm doing here. Some are pretty clear (docstring added on master right below a new definition in this PR – keep both), others were way over my head – e.g. the previous modification to `logdet` in `base/linalg/lu.jl`. cc: @simonbyrne, @KristofferC, @andreasnoack, @jiahao, @sarvjeetsinghghotra.

File: base/math.jl [hunk 2]
```diff
@@ -13,9 +13,8 @@ export sin, cos, tan, sinh, cosh, tanh, asin, acos, atan,
        cbrt, sqrt, erf, erfc, erfcx, erfi, dawson,
        significand,
        lgamma, hypot, gamma, lfact, max, min, minmax, ldexp, frexp,
-       clamp, clamp!, modf, ^, mod2pi,
-       airyai, airyaiprime, airybi, airybiprime,
-       airyaix, airyaiprimex, airybix, airybiprimex,
+       clamp, modf, ^, mod2pi, rem2pi,
+       airy, airyai, airyprime, airyaiprime, airybi, airybiprime, airyx,
```
Comment (tkelman) [inline base/math.jl]: wrong conflict resolution on the airy names I think

File: base/math.jl [hunk 3]
```diff
@@ -708,6 +713,35 @@ const pi3o2_l  = 1.8369701987210297e-16 # convert(Float64, pi * BigFloat(3/2) -
...
+        else # add pi/2
+            return add22condh(y[1],y[2],pi1o2_h,pi1o2_l)
+        end
+    end
+end
+
+rem2pi(x::Float32, ::RoundingMode{:Nearest}) = Float32(rem2pi(Float64(x),RoundNearest))
+rem2pi(x::Int32, ::RoundingMode{:Nearest}) = rem2pi(Float64(x),RoundNearest)
+function rem2pi(x::Int64, ::RoundingMode{:Nearest})
+  fx = Float64(x)
```
Comment (tkelman) [inline base/math.jl]: 4 space indent

File: base/exports.jl (hunk 1 above)
Comment (StefanKarpinski) [inline base/exports.jl]: :man_shrugging:

Comment (StefanKarpinski) [inline base/exports.jl]: Honestly, I'm not sure. I'm also not sure why `rem2pi` takes a rounding argument but `mod2pi` doesn't.

Comment (simonbyrne) [inline base/exports.jl]: `mod2pi(x)` is `rem2pi(x, RoundDown)`

**Comment by simonbyrne** [comment]
Updated, but still needs some tests.

File: base/math.jl [hunk 4]
```diff
@@ -620,6 +620,42 @@ function frexp{T<:AbstractFloat}(x::T)
...
+
+    x - y*round(x/y,r)
+
+without any intermediate rounding.
+
+- if `r == RoundNearest`, then the result is exact, and in the interval
+  `[-abs(y)/2, abs(y)/2]`.
+
+- if `r == RoundToZero` (default), then the result is exact, and in the interval
+  `[0,abs(y))` if `x` is positive, or `(abs(y),0]` otherwise.
```
Comment (tkelman) [inline base/math.jl]: maybe math-mode (double backtick) the intervals since that isn't Julia notation?

**Comment by simonbyrne** [comment]
Two remaining questions:
 - should we deprecate `mod2pi`?
 - should we give `rem2pi` a default rounding mode?

File: base/math.jl (hunk 4 above)
Comment (simonbyrne) [inline base/math.jl]: Good call.

File: base/math.jl [hunk 5]
```diff
@@ -709,63 +745,166 @@ const pi4o2_h  = 6.283185307179586      # convert(Float64, pi * BigFloat(2))
...
+Compute the remainder of `x` after integer division by `2π`, with the quotient rounded
+according to the rounding mode `r`. In other words, the quantity
 
-This function computes a floating point representation of the modulus after division by
-numerically exact `2π`, and is therefore not exactly the same as `mod(x,2π)`, which would
-compute the modulus of `x` relative to division by the floating-point number `2π`.
+    x - 2π*round(x/(2π),r)
+
+without any intermediate rounding. This internally uses a high precision approximation of
+2π, and so will give a more `rem(x,2π,r)`
```
Comment (tkelman) [inline base/math.jl]: a more ... accurate result than ?

File: base/math.jl [hunk 6]
```diff
@@ -709,63 +745,166 @@ const pi4o2_h  = 6.283185307179586      # convert(Float64, pi * BigFloat(2))
...
+
+- if `r == RoundNearest`, then the result is in the interval `[-π,π]`. This will generally
+  be the most accurate result.
+
+- if `r == RoundToZero`, then the result is in the interval `[0,2π]` if `x` is positive,.
+  or `[2π,0]` otherwise.
```
Comment (tkelman) [inline base/math.jl]: ``[-2π,0]`` ?

**Comment by StefanKarpinski** [comment]
Why is `rem2pi` preferable to `mod2pi`? Doesn't one need each in different situations? Shouldn't they both take rounding arguments and default to the current rounding mode?

File: base/math.jl [hunk 7]
```diff
@@ -709,63 +745,166 @@ const pi4o2_h  = 6.283185307179586      # convert(Float64, pi * BigFloat(2))
...
+This function computes a floating point representation of the modulus after division by
+numerically exact `2π`, and is therefore not exactly the same as `mod(x,2π)`, which would
+compute the modulus of `x` relative to division by the floating-point number `2π`.
+
+```jldoctest
+julia> mod2pi(9*pi/4)
+0.7853981633974481
+```
+"""
+mod2pi(x) = rem2pi(x,RoundDown)
```
Comment (simonbyrne) [inline base/math.jl]: @StefanKarpinski `mod2pi` is now just a special case of `rem2pi`

**Comment by simonbyrne** [comment]
I've added some more tests. Comments welcome.

**Comment by StefanKarpinski** [comment]
Can't we just make `rem2pi` a method of `mod2pi` then?

**Comment by simonbyrne** [comment]
I think `rem` is more general that `mod` (`mod` to me implies that the values are always in the interval `[0,x)` for some `x > 0`).

**Comment by tkelman** [comment]
rebase then merge IMO?

**Comment by tkelman** [comment]
might be worth adding a news entry sometime before the release

**Comment by StefanKarpinski** [comment]
That is not correct: mod means that the value is in [0,m) for positive m or (m,0] for negative m, I.e. that the result sign always matches the modulus; rem means that the sign matches the argument rather than the modulus. For mathematical purposes, one generally wants mod not rem.

**Comment by simonbyrne** [comment]
I took rem to mean remainder after division, as described in #9283. In that case mod is a special case (being when the division is rounded down).

**Comment by StefanKarpinski** [comment]
@simonbyrne and I just had the following chat conversation which cleared this issue up for me.

---

Stefan Karpinski [2:40 PM]
> @simonbyrne: I’m a bit unclear on what the difference between my understanding of `rem` and yours is

Simon Byrne [2:44 PM]
> I'm not sure I understand what you're proposing
> are you saying we should just make `rem(x, y, r::RoundingMode)`, `mod(x, y, r::RoundingMode)`?

Stefan Karpinski [2:59 PM]
> I guess I don’t understand why rem is more fundamental than mod in your definition of it
> @simonbyrne ^
> (sorry, trying to get this happening as a synchronous conversation)

Simon Byrne [3:00 PM]
> to me, `mod(x,y)` implies that the result must be in the interval `[0,y)` (if `y>0`)

Stefan Karpinski [3:01 PM]
> the more general definition is that mod(x,y) is in `[0,y)` for positive y and `(y,0]` for negative y

Simon Byrne [3:01 PM]
> sure

Stefan Karpinski [3:01 PM]
> which is usually what you actually want for a mathematical modulus operation

Simon Byrne [3:01 PM]
> yes
> but then what do you call the operation `r = f(x,y)` which gives `-abs(y)/2 <= x <= abs(y)/2`?
> I tried to explain my thinking here:
> https://github.com/JuliaLang/julia/issues/9283
> The idea being that `rem` and `mod` are both the same sort of thing
> finding `r` such that `x = q*y + r` for some integer `q`
> and the difference is how you compute `q`
> (or, really, the direction in which you round `q`)
> and so "remainder" seems like the most general term for this operation
> (similarly, we could also make `fld` and `cld` special cases of `div(x,y,::RoundingMode)`)

Stefan Karpinski [3:06 PM]
> Why make rem the fundamental one and not mod?

Simon Byrne [3:06 PM]
> because you're finding a remainder
> not a modulus

Stefan Karpinski [3:06 PM]
> Can you explain how rounding is the difference between rem and mod?

Simon Byrne [3:08 PM]
> so current `rem` is defined as
> `rem(x,y) = x - y*trunc(x/y)` (edited)
> (without intermediate rounding)
> `mod(x,y) = x - y*floor(x/y)` (edited)
> so #10946 defined
> `rem(x,y,r::RoundingMode) = x - y*round(x/y, r)`
> i.e. the remainder after Euclidean division with rounding direction `r`

Stefan Karpinski [3:11 PM]
> I see
> So you can get the operation you were talking about via a different rounding mode

Simon Byrne [3:12 PM]
> yes, that would just be
> `rem(x,y, RoundNearest) = x - y*round(x/y)`

Stefan Karpinski [3:13 PM]
> standard rem, mod correspond to round towards zero (trump), and round down (floor), respectively

Simon Byrne [3:13 PM]
> yes
> exactly

Stefan Karpinski [3:13 PM]
> Ok, I'm starting to get this now

Simon Byrne [3:14 PM]
> my plan is also to define corresponding methods for `div` as well (edited)
> since that is useful (e.g. for implementing fixed point arithmetic)

Stefan Karpinski [3:14 PM]
> So the question becomes what function to make the three-argument form a method of

Simon Byrne [3:15 PM]
> yes

Stefan Karpinski [3:15 PM]
> A new function entirely, rem or mod

Simon Byrne [3:15 PM]
> yes
> personally, remainder seems like the best name for it

Stefan Karpinski [3:16 PM]
> Is the issue with mod2pi that the behavior it implements does not correspond to what the mod function does?

Simon Byrne [3:17 PM]
> no
> our current `mod2pi(x) = mod(x, 2pi)`
> which is okay
> but not that useful

Stefan Karpinski [3:17 PM]
> But with better precision

Simon Byrne [3:17 PM]
> it's better to do the argument reduction to `[-pi, pi]`, rather than `[0, 2pi]`

Stefan Karpinski [3:18 PM]
> Ah, I see

Simon Byrne [3:18 PM]
> e.g. if you get `x = -0.000001`

Stefan Karpinski [3:19 PM]
> The trouble is that this operation has no standard name

Simon Byrne [3:19 PM]
> and so that corresponds to `rem(x, 2pi, RoundNearest)`

Stefan Karpinski [3:19 PM]
> Right, right

Simon Byrne [3:19 PM]
> hence `rem2pi(x, RoundNearest)`
> i mean we could punt on dispatch

Stefan Karpinski [3:19 PM]
> ?

Simon Byrne [3:19 PM]
> define an `twopi::Irrational`
> and `rem(x, twopi, RoundNearest)`

Stefan Karpinski [3:20 PM]
> I think you mean tau

Simon Byrne [3:20 PM]
> sure

Stefan Karpinski [3:21 PM]
> I did add tau in a PR at one point, but I think I'd generally rather not start dispatching on irrationals too much

Simon Byrne [3:21 PM]
> yeah

Stefan Karpinski [3:21 PM]
> Gives people the wrong idea

Simon Byrne [3:21 PM]
> it's not that useful
> and could trigger a lot of unnecessary bikeshedding
> best to leave that to `Tau.jl`

Stefan Karpinski [3:22 PM]
> If pi works, why doesn't 2pi, and 3pi, and pi/2, etc
> Sure

Simon Byrne [3:22 PM]
> yeah
> C calls `rem(x,y,RoundNearest)` `remainder` (edited)

Stefan Karpinski [3:23 PM]
> I could actually imagine extending irrational to having integer and rational multiples

Simon Byrne [3:23 PM]
> as does the IEEE spec

Stefan Karpinski [3:23 PM]
> But that's way too big a change to drag into this

Simon Byrne [3:23 PM]
> yeah
> agreed
> maybe for 2.0
> or 3.14 if we're going for TeX numbering

Stefan Karpinski [3:24 PM]
> Is that how rem is defined for floating point in C?

Simon Byrne [3:24 PM]
> no `rem` and `remainder` are different

Stefan Karpinski [3:24 PM]
> I.e. In a way that is incompatible with integer rem?
> Oh, I see. It calls that operation remainder

Simon Byrne [3:25 PM]
> C has `fmod` (which is `rem(x::Float64, y::Float64`)
> and `remainder` (which is `rem(x::Float64, y::Float64, RoundNearest)`)

Stefan Karpinski [3:26 PM]
> I wonder if we should run this by a few people with good judgement on this sort of thing, e.g. @jeffbezanson and @stevengj

Simon Byrne [3:27 PM]
> and `remquo`, which gives the same answer as `remainder`, but also the last few bits of the quotient via a pointer
> it doesn't define a `mod` function

Stefan Karpinski [3:27 PM]
> I can see rem being a good place for this but something feels oddly asymmetrical about it

Simon Byrne [3:27 PM]
> which is why we had all the hassles defining it back in the day
> (the day being 2013 or so)

Stefan Karpinski [3:28 PM]
> Ok, so your plan is to make rem/div the fundamental pair with three-arg methods

Simon Byrne [3:28 PM]
> yes

Stefan Karpinski [3:29 PM]
> And then conceptually make all the two-arg versions aliases for calling those with various rounding modes

Simon Byrne [3:29 PM]
> yes

Stefan Karpinski [3:29 PM]
> What about rem2pi(x)
> Which rounding more does that imply?

Simon Byrne [3:30 PM]
> the same as `rem(x, 2pi)`, i.e. `RoundToZero`

Stefan Karpinski [3:30 PM]
> Sign trunc, i.e. Sign matches x?

Simon Byrne [3:30 PM]
> yes

Stefan Karpinski [3:30 PM]
> And round to nearest doesn't get a name

Simon Byrne [3:30 PM]
> i mean, we could make it the default?
> but I think people would find that confusing

Stefan Karpinski [3:31 PM]
> Yes, that seems pretty confusing

Simon Byrne [3:31 PM]
> also would break a lot

Stefan Karpinski [3:31 PM]
> Ok, thanks for explaining this to me
> You've sold me on your grand vision

**Pull request merged**